TLM/TLM/Patch/_VehicleAI/_TrainAI/CheckNextLanePatch.cs

```diff
@@ -0,0 +1,230 @@
+        private delegate void TargetDelegate(ushort vehicleID,
+                                             ref Vehicle vehicleData,
+                                             ref float maxSpeed,
+                                             PathUnit.Position position,
+                                             uint laneID,
+                                             byte offset,
+                                             PathUnit.Position prevPos,
+                                             uint prevLaneID,
+                                             byte prevOffset,
+                                             Bezier3 bezier);
+
+        [UsedImplicitly]
+        public static MethodBase TargetMethod() => TranspilerUtil.DeclaredMethod<TargetDelegate>(typeof(TrainAI), "CheckNextLane");
+
+        private static CheckOverlapDelegate CheckOverlap;
+
+        [UsedImplicitly]
+        public static void Prepare() {
+            CheckOverlap = Constants.ManagerFactory.GameConnectionManager.TrainAIConnection.CheckOverlap;
+        }
+
+        [UsedImplicitly]
+        public static bool Prefix(TrainAI __instance,
+                                  ushort vehicleID,
+                                  ref Vehicle vehicleData,
+                                  ref float maxSpeed,
+                                  PathUnit.Position position,
+                                  uint laneID,
+                                  byte offset,
+                                  PathUnit.Position prevPos,
+                                  uint prevLaneID,
+                                  byte prevOffset,
+                                  Bezier3 bezier) {
+            NetManager netManager = NetManager.instance;
+
+            ushort nextSourceNodeId = offset < position.m_offset
+                                          ? netManager.m_segments.m_buffer[position.m_segment].m_startNode
+                                          : netManager.m_segments.m_buffer[position.m_segment].m_endNode;
+
+            ushort refTargetNodeId = prevOffset == 0
+                                         ? netManager.m_segments.m_buffer[prevPos.m_segment].m_startNode
+                                         : netManager.m_segments.m_buffer[prevPos.m_segment].m_endNode;
+
+#if DEBUG
+            bool logLogic = DebugSwitch.CalculateSegmentPosition.Get()
+                           && (DebugSettings.NodeId <= 0
+                               || refTargetNodeId == DebugSettings.NodeId)
+                           && (GlobalConfig.Instance.Debug.ApiExtVehicleType == ExtVehicleType.None
+                               || GlobalConfig.Instance.Debug.ApiExtVehicleType == ExtVehicleType.RailVehicle)
+                           && (DebugSettings.VehicleId == 0
+                               || DebugSettings.VehicleId == vehicleID);
+
+            if (logLogic) {
+                Log._Debug($"CustomTrainAI.CustomCheckNextLane({vehicleID}) called.\n" +
```

Comment:
multiple calls to String.Format, luckily only if logLogic is true

Reply:
Not much I can do. Usually used then debugging specific node and vehicle type - I used it only once.
One way to "solve it" is conversion to interpolated string but then it must be single line invocation - 500 chars, bad idea 😄 
I'll leave it like it is, because it's not the only place where similar problem exist.